File: scripts/generate_icons.mjs
/**
 * Generate app icons for Windows and macOS from SVG logo.
 * macOS icons require ~12.8% padding around the icon content (Apple HIG).
 */

import sharp from 'sharp';
import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const srcTauri = path.join(__dirname, '..', 'src-tauri');
const iconsDir = path.join(srcTauri, 'icons');
const svgPath = path.join(srcTauri, 'logo.svg');

// Ensure icons directory exists
if (!fs.existsSync(iconsDir)) {
    fs.mkdirSync(iconsDir, { recursive: true });
}

console.log('='.repeat(50));
console.log('POI Collector Icon Generator');
console.log('='.repeat(50));

// Step 1: Create base 1024x1024 PNG from SVG
console.log('\n[1/3] Converting SVG to PNG...');
const basePng = await sharp(svgPath)
    .resize(1024, 1024)
    .png()
    .toBuffer();

// Save base icon (for Windows - no padding needed)
await sharp(basePng).toFile(path.join(iconsDir, 'icon.png'));
console.log('  Created: icon.png (1024x1024, no padding)');

// Step 2: Create macOS version with 12.8% padding
console.log('\n[2/3] Creating macOS icon with 12.8% padding...');
const paddingPercent = 12.8;
const scaleFactor = (100 - 2 * paddingPercent) / 100; // ~0.744
const newIconSize = Math.floor(1024 * scaleFactor); // ~762
const offset = Math.floor((1024 - newIconSize) / 2); // ~131

// Resize icon content smaller
const scaledIcon = await sharp(basePng)
    .resize(newIconSize, newIconSize, { fit: 'contain' })
    .toBuffer();

// Create transparent canvas and composite the scaled icon centered
const macosIcon = await sharp({
    create: {
        width: 1024,
        height: 1024,
        channels: 4,
        background: { r: 0, g: 0, b: 0, alpha: 0 }
    }
})
    .composite([{
        input: scaledIcon,
        left: offset,
        top: offset
    }])
    .png()
    .toFile(path.join(iconsDir, 'icon_macos.png'));

console.log(`  Created: icon_macos.png (1024x1024, with ${paddingPercent}% padding)`);

// Step 3: Generate all required icon sizes
console.log('\n[3/3] Generating icon sizes...');

// Standard sizes for Tauri
const sizes = [16, 24, 32, 48, 64, 128, 256, 512, 1024];
for (const size of sizes) {
    await sharp(basePng)
        .resize(size, size)
        .png()
        .toFile(path.join(iconsDir, `${size}x${size}.png`));
    console.log(`  Created: ${size}x${size}.png`);
}

// Tauri special sizes
await sharp(basePng).resize(128, 128).png().toFile(path.join(iconsDir, '128x128.png'));
await sharp(basePng).resize(256, 256).png().toFile(path.join(iconsDir, '128x128@2x.png'));
console.log('  Created: 128x128.png and 128x128@2x.png');

// Windows Store logos
const storeLogos = {
    'Square30x30Logo.png': 30,
    'Square44x44Logo.png': 44,
    'Square71x71Logo.png': 71,
    'Square89x89Logo.png': 89,
    'Square107x107Logo.png': 107,
    'Square142x142Logo.png': 142,
    'Square150x150Logo.png': 150,
    'Square284x284Logo.png': 284,
    'Square310x310Logo.png': 310,
    'StoreLogo.png': 50,
};

for (const [name, size] of Object.entries(storeLogos)) {
    await sharp(basePng)
        .resize(size, size)
        .png()
        .toFile(path.join(iconsDir, name));
    console.log(`  Created: ${name}`);
}

console.log('\n' + '='.repeat(50));
console.log('PNG icons generated successfully!');
console.log('='.repeat(50));
console.log(`\nGenerated files in: ${iconsDir}`);
console.log('\nNext: Running Tauri icon command to generate .ico and .icns...');

File: scripts/generate_icons.py
#!/usr/bin/env python3
"""
Generate app icons for Windows and macOS from SVG logo.
macOS icons require ~12.8% padding around the icon content.
"""

import subprocess
import sys
from pathlib import Path

# Check for required packages
try:
    from PIL import Image
    import cairosvg
except ImportError as e:
    print(f"Missing required package: {e}")
    print("Installing required packages...")
    subprocess.check_call([sys.executable, "-m", "pip", "install", "pillow", "cairosvg"])
    from PIL import Image
    import cairosvg

def svg_to_png(svg_path: Path, png_path: Path, size: int = 1024):
    """Convert SVG to PNG at specified size."""
    cairosvg.svg2png(
        url=str(svg_path),
        write_to=str(png_path),
        output_width=size,
        output_height=size
    )
    print(f"Created: {png_path}")

def add_padding(input_path: Path, output_path: Path, padding_percent: float = 12.8):
    """
    Add padding around the icon for macOS.
    Apple HIG recommends icon content be within ~80% of the canvas.
    This adds padding_percent% padding on each side.
    """
    img = Image.open(input_path)
    original_size = img.size[0]  # Assuming square
    
    # Calculate new size after scaling down
    scale_factor = (100 - 2 * padding_percent) / 100  # ~0.744 for 12.8%
    new_icon_size = int(original_size * scale_factor)
    
    # Resize the icon
    resized_icon = img.resize((new_icon_size, new_icon_size), Image.Resampling.LANCZOS)
    
    # Create new canvas with transparent background
    canvas = Image.new('RGBA', (original_size, original_size), (0, 0, 0, 0))
    
    # Calculate position to center the icon
    offset = (original_size - new_icon_size) // 2
    
    # Paste resized icon onto canvas
    canvas.paste(resized_icon, (offset, offset))
    
    canvas.save(output_path, 'PNG')
    print(f"Created with {padding_percent}% padding: {output_path}")

def main():
    # Paths
    project_root = Path(__file__).parent.parent
    src_tauri = project_root / "src-tauri"
    icons_dir = src_tauri / "icons"
    svg_path = src_tauri / "logo.svg"
    
    # Ensure icons directory exists
    icons_dir.mkdir(exist_ok=True)
    
    # Temp paths
    temp_png = icons_dir / "temp_1024.png"
    windows_icon_source = icons_dir / "icon.png"
    macos_icon_source = icons_dir / "icon_macos.png"
    
    print("=" * 50)
    print("POI Collector Icon Generator")
    print("=" * 50)
    
    # Step 1: Convert SVG to PNG (1024x1024)
    print("\n[1/4] Converting SVG to PNG...")
    svg_to_png(svg_path, temp_png, 1024)
    
    # Step 2: Create Windows icon source (no padding needed for Windows)
    print("\n[2/4] Creating Windows icon source...")
    img = Image.open(temp_png)
    img.save(windows_icon_source, 'PNG')
    print(f"Created: {windows_icon_source}")
    
    # Step 3: Create macOS icon source with 12.8% padding
    print("\n[3/4] Creating macOS icon source with 12.8% padding...")
    add_padding(temp_png, macos_icon_source, padding_percent=12.8)
    
    # Step 4: Generate all required icon sizes
    print("\n[4/4] Generating icon sizes...")
    
    # Windows/general icons (from non-padded source)
    windows_sizes = [16, 24, 32, 48, 64, 128, 256, 512, 1024]
    for size in windows_sizes:
        img = Image.open(windows_icon_source)
        resized = img.resize((size, size), Image.Resampling.LANCZOS)
        resized.save(icons_dir / f"{size}x{size}.png", 'PNG')
        print(f"  Created: {size}x{size}.png")
    
    # Special Tauri sizes
    img = Image.open(windows_icon_source)
    img.resize((128, 128), Image.Resampling.LANCZOS).save(icons_dir / "128x128.png", 'PNG')
    img.resize((256, 256), Image.Resampling.LANCZOS).save(icons_dir / "128x128@2x.png", 'PNG')
    print("  Created: 128x128.png and 128x128@2x.png")
    
    # Windows Store logos
    store_sizes = {
        "Square30x30Logo.png": 30,
        "Square44x44Logo.png": 44,
        "Square71x71Logo.png": 71,
        "Square89x89Logo.png": 89,
        "Square107x107Logo.png": 107,
        "Square142x142Logo.png": 142,
        "Square150x150Logo.png": 150,
        "Square284x284Logo.png": 284,
        "Square310x310Logo.png": 310,
        "StoreLogo.png": 50,
    }
    for name, size in store_sizes.items():
        resized = img.resize((size, size), Image.Resampling.LANCZOS)
        resized.save(icons_dir / name, 'PNG')
        print(f"  Created: {name}")
    
    # Clean up temp file
    temp_png.unlink()
    
    print("\n" + "=" * 50)
    print("Icon generation complete!")
    print("=" * 50)
    print(f"\nGenerated files in: {icons_dir}")
    print("\nNext steps:")
    print("  1. Use 'npx tauri icon' to generate .ico and .icns files")
    print("     - For Windows: npx tauri icon src-tauri/icons/icon.png")
    print("     - For macOS: Use icon_macos.png as source for .icns")
    print("\nNote: macOS icon has 12.8% padding per Apple HIG standards")

if __name__ == "__main__":
    main()
